# Import Libraries

In [1]:
#import libraries
import pandas as pd
import numpy as np
import os

# Import Data

In [2]:
path = r"C:\Users\bhimc\Documents\Instacart Basket Analysis (Career Foundry Rekha)"

In [3]:
path

'C:\\Users\\bhimc\\Documents\\Instacart Basket Analysis (Career Foundry Rekha)'

In [17]:
#import ords_prods_merged dataframe
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

# 01 User-Defined Functions

In [18]:
#Create a subset of the ords_prods_merged dataframe for the first million rows
df = ords_prods_merge[:1000000]

In [6]:
df.shape

(1000000, 15)

In [7]:
# Define a function

def price_label(row):
    
    if row['prices'] <= 5:
        return 'Low range product'
    elif (row['prices'] > 5) and (row['prices'] <= 15):
        return 'Mid range product'
    elif row['prices'] > 15:
        return 'High range product'
    else: return np.nan
    

In [9]:
#Apply the function

df['prices_range'] = df.apply(price_label, axis=1)

C:\Users\bhimc\AppData\Local\Temp\ipykernel_9700\1212145324.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['prices_range'] = df.apply(price_label, axis=1)


In [10]:
df['prices_range'].value_counts(dropna = False)

prices_range
Mid range product    756450
Low range product    243550
Name: count, dtype: int64

In [11]:
#Check what the most expensive product within the subset
df['prices'].max()

14.8

# 02 If-Statements with the loc() Function

In [12]:
#Create conditions

In [13]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High range product'

C:\Users\bhimc\AppData\Local\Temp\ipykernel_9700\2555204669.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High range product'


In [14]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid range product'

In [15]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low range product'

In [16]:
df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid range product    756450
Low range product    243550
Name: count, dtype: int64

In [19]:
ords_prods_merge.loc[ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [21]:
ords_prods_merge.loc[(ords_prods_merge['prices'] <= 15) & (ords_prods_merge['prices'] > 5), 'price_range_loc'] =  'Mid_range product'

In [22]:
ords_prods_merge.loc[ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [23]:
ords_prods_merge['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid_range product     21860860
Low-range product     10126321
High-range product      417678
Name: count, dtype: int64

# 03 If-Statements with For-Loops

In [24]:
ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [25]:
result = []

for value in ords_prods_merge["orders_day_of_week"]:
    if value == 0:
        result.append("Busiest day")
    elif value == 4:
        result.append("Least busy")
    else: 
        result.append("Regularly busy")

In [26]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [27]:
#create a new column
ords_prods_merge['busiest_day'] = result

In [28]:
ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

# 04 Task 4.7

## Step 2

In [29]:
# Create column busiest_day to busiest day, identify the two busiest days of the week (0,1), and the two slowest days of the week (3,4), or regularly busy(all other days)

result_2 = []

for value in ords_prods_merge["orders_day_of_week"]:
    if value == 0 or value ==1:
        result_2.append("Busiest days")
    elif value == 3 or value == 4:
        result_2.append("Slowest days")
    else:
        result_2.append("Regularly busy")
        
    

In [30]:
result_2

['Regularly busy',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Slowest days',
 'Regularly busy',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Slowest days',
 'Slowest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest 

## Step 3

In [31]:
# "Check the values of this new column for accuracy. Note any observations in markdown format"

In [32]:
ords_prods_merge['busiest_days'] = result_2

In [33]:
ords_prods_merge['busiest_days'].value_counts(dropna = False)

busiest_days
Regularly busy    12916111
Busiest days      11864412
Slowest days       7624336
Name: count, dtype: int64

In [34]:
ords_prods_merge.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered product_name  aisle_id  department_id  prices _merge  \
0          0         Soda        77              7     9.0   both   
1          1         Soda        77              7     9.0   both   
2          1         Soda        77              7     9.0   both   
3          1         Soda        77              7     9.0   both   
4          1         Soda        77              7     9.0   both   

     price_range_loc     busiest_day    busiest_days  
0  Mid_range product  Regularly busy  Regularly busy  
1  Mid_range product  Regularly busy    Slowest days  
2  Mid_range product  Regularly busy    Slowest days  
3  Mid_range product      Least busy    Slowest days  
4  Mid_range product      Least busy    Slowest days

### The slowest days is equal to 7624336, and it matches orders_day_of_week (3 and 4) because 3840534 + 3783802 = 7624336. The same also applies to the busiest days (0 and 1) and the regular days (all other numbers). 

## Step 4

In [35]:
# "The senior technical officer at Instacart wants you to identify the busiest hours of the day."

In [36]:
ords_prods_merge['order_hour_of_day'].value_counts()

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

In [37]:
# Since there are 24 hours, the hours can be grouped in thirds. 
#The "most orders" can be grouped as 10,11,14,15,13,12,16
# The "Fewest orders" can be grouped as 3,4,2,5,1,0,6,23
# The "Average orders" can be grouped with the remaining amount of numbers


In [39]:
result_3 = []

for value in ords_prods_merge["order_hour_of_day"]:
    if value in [10,11,14,15,13,12,16]:
        result_3.append("Most orders")
    elif value in [3,4,2,5,1,0,6,23]:
        result_3.append("Fewest orders")
    else:
        result_3.append("Average orders")

In [40]:
result_3

['Average orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Most orders',

In [41]:
# new column containing those labels
ords_prods_merge['busiest_period_of_day'] = result_3

In [42]:
ords_prods_merge.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered product_name  aisle_id  department_id  prices _merge  \
0          0         Soda        77              7     9.0   both   
1          1         Soda        77              7     9.0   both   
2          1         Soda        77              7     9.0   both   
3          1         Soda        77              7     9.0   both   
4          1         Soda        77              7     9.0   both   

     price_range_loc     busiest_day    busiest_days busiest_period_of_day  
0  Mid_range product  Regularly busy  Regularly busy        Average orders  
1  Mid_range product  Regularly busy    Slowest days        Average orders  
2  Mid_range product  Regularly busy    Slowest days           Most orders  
3  Mid_range product      Least busy    Slowest days        Average orders  
4  Mid_range product      Least busy    Slowest days           Most orders

## Step 5

In [43]:
# "Print the frequency for this new column"

In [44]:
ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most orders       18663868
Average orders    12451854
Fewest orders      1289137
Name: count, dtype: int64

## Step 7

In [47]:
# "Export your dataframe as a pickle file (since you added new columns) and store it correctly in your “Prepared Data” folder."

In [46]:
ords_prods_merge.to_pickle(os.path.join(path, '02 data', 'Prepared data', 'orders_products_merged_derived.pkl'))